### PR TITLE
bigdft: use -fPIC when building shared library

### PR DESCRIPTION
**Pull request by iamashwin99**
The default behavior of bigdft-suite was updated to build as a shared library in https://github.com/spack/spack/pull/41562 . 
This however broke bigdft-spred as bigdft-suite doesn't compile with `-fPIC` by default and hence we need to add those compiler flags manually when building the shared library variant. I am not sure if this is something to be reported as a "bug" in the upstream package ( ie. build system design should be intelligent to do this on its own but their package doesn't). Im at the moment trying to fix that issue by manually adding the right flags when a shared library is built. 

The MR is in a draft state as it is still not compiling without errors and I need some time to solve that.

### CODE REVIEW

**Review by bernhardkaindl — APPROVED** [review]
@iamashwin99 I give you a tentative approval because it looks good. Still, if you have no further changes (e.g. also no changes to commit messages etc to give some explanation), then after pushing your last change, you should click the button labelled "Ready for Review":

- This should then cause GitLab CI to be unblocked to run it automatically (we need to pass it for all merges - no way around it for us ATM)
- And after that has passed, allow us to merge.

Here are further comments that I also added to the PR's description (which you can fill out, just replace the questions that you answer with your answers)
@iamashwin99: Could you fill out the description of this PR here with some information:

- Why is this PR still set as Draft (not ready for merge)?
  - Do you need to merge more changes?
  - Or were you unable to run your tests?

- Once clear, could you click the button labelled "Ready for review" below?
   It is to the right of:
    > This pull request is still a work in progress
    > Draft pull requests cannot be merged.

You should see that GitLab CI has not started because it is still a draft PR.

So if you don't want to make further changes (you can also skip the `git rebase develop` if all looks good, then clicking the "Ready for Review" button should start the GitLab CI (eventually), and we need GitLab CI for merging as well.

Maybe obsolete: Did you report this to the upstream package managers, are they aware of it and are fixing the newer versions when they get `--enable-dynamic-libraries` like `spack` enable it?
```py
        if spec.satisfies("+shared"):
            args.append("--enable-dynamic-libraries")
```
After everything is clear, could you then pull the main `spack` repo's `develop` branch to your local repo and then rebase your changes using `git rebase develop`?

Then, the git history before your commits will include the latest commits on the `spack` master branch.

Because this PR is from January, its base is outdated and you should rebase the latest `develop`.

When all is correct,  push again with `git push --force-with-lease`?

After the latest commit on the `develop` that you rebased on has passed CI, GitLab CI should run, and you need to set it to ready for review for this to happen eventually.

-- @bernhardkaindl

**Review by bernhardkaindl — APPROVED** [review]
@danielecesarini It seems GitHub assumes you changed or added some of these build recipes.
At least, GitHub suggested requesting a review from you, which I've done now.

Could you give us some review comment on this PR within a week (or at least 7 days)?

Thanks!

**Comment by bernhardkaindl** [comment]
@iamashwin99 Could you ping (notify us by commenting with @OurGitHubHandles (just use those that are suggested to you when you type the `@`

- After having filled out at least some infos for the PR description
- and after having given the "Ready for Review"
- and after GitLab PR passed
- and a week or 7 days passed or so:
  - notify us again of the status?

If you don't know whom to inform (lets say by 29 January the latest anyone, but just lets say me or another regular reviewers of PRs), start to notify those reviewers / maintainers of your PR then?

**Comment by iamashwin99** [comment]
Thanks for the reminder, I was trying to fix build errors caused in the shared variant of the bigdft-suite.
I left the MR in a draft state as I still had difficulties compiling the package `bigdft-spread` and so the MR is not in a mergeable state. I wanted to have an MR here anyways, to indicate that I am working on this issue. Sorry for not updating the MR in detail.

**Comment by danielecesarini** [comment]
> @danielecesarini It seems GitHub assumes you changed or added some of these build recipes. At least, GitHub suggested requesting a review from you, which I've done now.
> 
> Could you give us some review comment on this PR within a week (or at least 7 days)?
> 
> Thanks!

The PR looks good to me

**Comment by iamashwin99** [comment]
I am currently trying to debug build failure in `bigdft-core` from my branch:
```
$SPACK_OPT/gcc-11.3.0/openmpi-5.0.1-lwubxivxecdwcqrmszjzolt25eczdjpt/bin/mpif90 -I. -I../includes -I./wavelib -I$SPACK_OPT/gcc-11.3.0/libxc-4.3.4-3wfpmtrelpvhpwvfse6r3u4enh56isjh/include -I$SPACK_OPT/gcc-11.3.0/bigdft-futile-1.9.2-bmiyvez3v33lfzpjfpkfupyf42mq2zch/include -fopenmp -fPIC -fallow-argument-mismatch -Wl,-rpath=$SPACK_OPT/gcc-11.3.0/bigdft-core-1.9.2-3ycvl66vhnuwrqqlio7vvu6rn246f5xp/lib  -o utilities utilities.o libbigdft-1.so.9 -labinit -L$SPACK_OPT/gcc-11.3.0/libxc-4.3.4-3wfpmtrelpvhpwvfse6r3u4enh56isjh/lib -lxc -L$SPACK_OPT/gcc-11.3.0/libxc-4.3.4-3wfpmtrelpvhpwvfse6r3u4enh56isjh/lib -lxcf90          -lCheSS-1 -lPSolver-1 -L$SPACK_OPT/gcc-11.3.0/bigdft-atlab-1.9.2-ggyqlhoxmy5wuswej24exfd7kw6zqqrh/lib -latlab-1 -L$SPACK_OPT/gcc-11.3.0/bigdft-futile-1.9.2-bmiyvez3v33lfzpjfpkfupyf42mq2zch/lib -lfutile-1   -L$SPACK_OPT/gcc-11.3.0/netlib-scalapack-2.2.0-qsmj453rryhrbvzgsyb27go6rjz2t7vv/lib -lscalapack -L$SPACK_OPT/gcc-11.3.0/openblas-0.3.25-qnjwp32cronlfhki7fllbu6biotz3hlp/lib -lopenblas -L$SPACK_OPT/gcc-11.3.0/openblas-0.3.25-qnjwp32cronlfhki7fllbu6biotz3hlp/lib -lopenblas       -L$SPACK_OPT/gcc-11.3.0/bigdft-chess-1.9.2-yecovahzlcgr4orz3ji6wj4uoybhvq53/lib -lCheSS-1 -L$SPACK_OPT/gcc-11.3.0/bigdft-futile-1.9.2-bmiyvez3v33lfzpjfpkfupyf42mq2zch/lib -lfutile-1 
$SPACK_OPT/gcc-11.3.0/openmpi-5.0.1-lwubxivxecdwcqrmszjzolt25eczdjpt/bin/mpif90 -I. -I../includes -I./wavelib -I$SPACK_OPT/gcc-11.3.0/libxc-4.3.4-3wfpmtrelpvhpwvfse6r3u4enh56isjh/include -I$SPACK_OPT/gcc-11.3.0/bigdft-futile-1.9.2-bmiyvez3v33lfzpjfpkfupyf42mq2zch/include -fopenmp -fPIC -fallow-argument-mismatch -Wl,-rpath=$SPACK_OPT/gcc-11.3.0/bigdft-core-1.9.2-3ycvl66vhnuwrqqlio7vvu6rn246f5xp/lib  -o test_forces test_forces.o libbigdft-1.so.9 -labinit -L$SPACK_OPT/gcc-11.3.0/libxc-4.3.4-3wfpmtrelpvhpwvfse6r3u4enh56isjh/lib -lxc -L$SPACK_OPT/gcc-11.3.0/libxc-4.3.4-3wfpmtrelpvhpwvfse6r3u4enh56isjh/lib -lxcf90          -lCheSS-1 -lPSolver-1 -L$SPACK_OPT/gcc-11.3.0/bigdft-atlab-1.9.2-ggyqlhoxmy5wuswej24exfd7kw6zqqrh/lib -latlab-1 -L$SPACK_OPT/gcc-11.3.0/bigdft-futile-1.9.2-bmiyvez3v33lfzpjfpkfupyf42mq2zch/lib -lfutile-1   -L$SPACK_OPT/gcc-11.3.0/netlib-scalapack-2.2.0-qsmj453rryhrbvzgsyb27go6rjz2t7vv/lib -lscalapack -L$SPACK_OPT/gcc-11.3.0/openblas-0.3.25-qnjwp32cronlfhki7fllbu6biotz3hlp/lib -lopenblas -L$SPACK_OPT/gcc-11.3.0/openblas-0.3.25-qnjwp32cronlfhki7fllbu6biotz3hlp/lib -lopenblas       -L$SPACK_OPT/gcc-11.3.0/bigdft-futile-1.9.2-bmiyvez3v33lfzpjfpkfupyf42mq2zch/lib -lfutile-1 
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_overlap_c_'
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_laplacian_c_'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:1025: bigdft] Error 1
make[3]: *** Waiting for unfinished jobs....
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_overlap_c_'
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_laplacian_c_'
collect2: error: ld returned 1 exit status
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `make[3]: *** [Makefile:1041: utilities] Error 1
c_overlap_c_'
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_overlap_c_'
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_laplacian_c_'
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_laplacian_c_'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:1029: frequencies] Error 1
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:1037: test_forces] Error 1
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_overlap_c_'
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_laplacian_c_'
collect2: error: ld returned 1 exit status
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_overlap_c_'
make[3]: *** [Makefile:1021: abscalc] Error 1
/usr/bin/ld: libbigdft-1.so.9: undefined reference to `c_laplacian_c_'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:1033: memguess] Error 1
make[3]: Leaving directory '$SPACK_stage/spack-stage-bigdft-core-1.9.2-3ycvl66vhnuwrqqlio7vvu6rn246f5xp/spack-src/bigdft/src'
make[2]: *** [Makefile:1506: all-recursive] Error 1
make[2]: Leaving directory '$SPACK_stage/spack-stage-bigdft-core-1.9.2-3ycvl66vhnuwrqqlio7vvu6rn246f5xp/spack-src/bigdft/src'
make[1]: *** [Makefile:634: all-recursive] Error 1
make[1]: Leaving directory '$SPACK_stage/spack-stage-bigdft-core-1.9.2-3ycvl66vhnuwrqqlio7vvu6rn246f5xp/spack-src/bigdft'
make: *** [Makefile:511: all] Error 2
```

**Comment by alalazo** [comment]
What is the status of this draft PR? Can we close it?

**Comment by iamashwin99** [comment]
I had issues compiling it before and I haven't had the time to fix it. Perhaps we can close it for now.

**Pull request closed without merging**
